traitsui/qt4/video_editor.py

```diff
@@ -0,0 +1,362 @@
+
+    def supportedPixelFormats(self, handleType):
+        return [QVideoFrame.Format_RGB32]
+
+    def present(self, frame):
+        cloned_frame = QVideoFrame(frame)
+        cloned_frame.map(QAbstractVideoBuffer.ReadOnly)
+        image = QImage(
+            cloned_frame.bits(),
+            cloned_frame.width(),
+            cloned_frame.height(),
+            cloned_frame.bytesPerLine(),
+            QVideoFrame.imageFormatFromPixelFormat(cloned_frame.pixelFormat())
+        )
+        self.frameAvailable.emit(image)
+        return True
+
+
+class VideoEditor(Editor):
+    """Traits UI 'display only' video editor.
+
+    This editor uses the Qt QMultimedia machinery to display video streams
+    to the screen.  Rather than being self-contained, the editor only concerns
+    itself with displaying the video, and provides traits that control
+    behaviour and provide internal state of the control during playback.
+    """
+
+    #: function to apply to the image. Takes frame to new frame
+    image_fun = Callable()
+
+    #: current held frame of the video viewer
+    frame = Property(Array(shape=(None, None, 3), dtype='u1'))
+
+    #: does the drawing onto the image plane
+    control = Instance(QVideoWidget)
+
+    #: handels the image pulling so the frames can be processed.
+    surface = Instance(QAbstractVideoSurface)
+
+    #: The QMediaObject that holds the connection to the video stream.
+    media_content = Instance(QMediaContent)
+
+    #: The QMediaPlayer that controls playback of the video stream.
+    media_player = Instance(QMediaPlayer)
+
+    #: The aspect ratio of the video editor.
+    aspect_ratio = AspectRatio()
+
+    #: The current state of the player, synchronized to the trait named
+    #: by factory.state.
+    state = PlayerState()
+
+    #: The current playback position of the player, synchronized to the trait
+    #: named by factory.position.
+    position = Float()
+
+    #: The total duration of the current video, synchronized to the trait
+    #: named by factory.duration.
+    duration = Float()
+
+    #: The media player playback status (loading, buffering, etc.).
+    media_status = MediaStatus()
+
+    #: The percentage of the buffer currently filled.
+    buffer = Range(0, 100)
+
+    #: A string holding the video error state, or "" if no error.
+    video_error = Str()
+
+    #: Whether the audio is muted or not.
+    muted = Bool(False)
+
+    #: The playback volume on a logarithmic scale (perceptually linear).
+    volume = Range(0.0, 100.0)
+
+    #: The playback rate.  Negative values rewind the video.
+    playback_rate = Float(1.0)
+
+    def _get_frame(self):
+        return self.control._np_image
+
+    def update_to_regular(self):
+        self.control = QVideoWidget()
+        self.control.setSizePolicy(
+            QSizePolicy.Expanding, QSizePolicy.Expanding
+        )
+        self.control.setBackgroundRole(QPalette.Window)
+        self.media_player.setVideoOutput(self.control)
```

Comment:
Not sure - it was a year ago that I wrote this.  There was some hackery needed to get display working properly on high-dpi Mac displays, so this might be part of that.